x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/action/EsqlResponseListener.java

```diff
@@ -69,8 +78,28 @@ public RestResponse buildResponse(EsqlQueryResponse esqlResponse) throws Excepti
                 ChunkedRestResponseBody.fromXContent(esqlResponse, channel.request(), channel)
             );
         }
+
         restResponse.addHeader(HEADER_NAME_TOOK_NANOS, Long.toString(System.nanoTime() - startNanos));
 
         return restResponse;
     }
+
+    /**
+     * Log the execution time and query when handling an ES|QL response.
+     */
+    public ActionListener<EsqlQueryResponse> wrapWithLogging() {
+        // We need to measure the execution time after handling the response/failure for the measurement to be correct.
+        return ActionListener.wrap(r -> {
+            onResponse(r);
+            LOGGER.info("Successfully executed ESQL query in {}ms:\n{}", timeMillis(), esqlQuery);
```

Comment:
I believe the execution time should exclude the time it takes to send the REST response. Hence, we should log before onResponse() or onFailure().

Reply:
Logging before `onFailure` is probably fine, but logging before `onResponse` will exclude the time to create the actual REST response (which could take a while depending on result set size).

I can, however, make it so that we log the same time that we also send back to the user [in the response header](https://github.com/elastic/elasticsearch/blob/386103f7b8ba8b454b349b057cad7205434de457/x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/action/EsqlResponseListener.java#L82).

I'd like to create a follow-up PR for this to add the more accurate logging in a clean way. (I could just move the success logging to the same location where the response time header is set, but I'd prefer to keep the success/failure logging in one place since the functionalities are related.)